src/main/kotlin/com/cognifide/gradle/aem/common/instance/service/repository/Node.kt

```diff
@@ -0,0 +1,210 @@
+import net.minidev.json.JSONArray
+import org.apache.jackrabbit.vault.util.JcrConstants
+
+class Node(private val repository: Repository, val path: String) : Serializable {
+
+    val name: String
+        get() = path.substringAfterLast("/")
+
+    private var propertiesLoaded: Properties? = null
+
+    /**
+     * Node properties.
+     *
+     * Keep in mind that these values are loaded lazily and sometimes it is needed to reload them
+     * using dedicated method.
+     */
+    val properties: Properties
+        get() = propertiesLoaded ?: reloadProperties()
+
+    /**
+     * JCR primary type of node.
+     */
+    @get:JsonIgnore
+    val type: String
+        get() = properties.string(JcrConstants.JCR_PRIMARYTYPE)!!
+
+    /**
+     * Parent node.
+     */
+    @get:JsonIgnore
+    val parent: Node
+        get() = Node(repository, path.substringBeforeLast("/"))
+
+    /**
+     * Get all node child nodes.
+     *
+     * Because of performance issues, using method is more preferred.
+     */
+    @get:JsonIgnore
+    val children: List<Node>
+        get() = children().toList()
+
+    /**
+     * Loop over all node child nodes.
+     */
+    @Suppress("unchecked_cast")
+    fun children(): Sequence<Node> = repository.http.get("$path.harray.1.json") { asJson(it) }
+            .run {
+                try {
+                    read<JSONArray>(Property.CHILDREN.value)
+                } catch (e: PathNotFoundException) {
+                    JSONArray() // no children
+                }
+            }
+            .map { child -> child as Map<String, Any> }
+            .map { props ->
+                Node(repository, "$path/${props[Property.NAME.value]}").apply {
+                    propertiesLoaded = Properties(filterMetaProperties(props))
+                }
+            }
+            .asSequence()
+
+    /**
+     * Create or update node in repository.
+     */
+    fun save(properties: Map<String, Any?>): RepositoryResult = try {
+        repository.http.postMultipart(path, postProperties(properties) + operationProperties("")) {
+            asObjectFromJson(it, RepositoryResult::class.java)
+        }
+    } catch (e: RequestException) {
+        throw RepositoryException("Cannot save repository node: $path", e)
+    }
+
+    /**
+     * Delete node and all children from repository.
+     */
+    fun delete(): RepositoryResult = try {
+        repository.http.postMultipart(path, operationProperties("delete")) {
+            asObjectFromJson(it, RepositoryResult::class.java)
+        }
+    } catch (e: RequestException) {
+        throw RepositoryException("Cannot delete repository node: $path", e)
+    }
+
+    /**
+     * Deletes node and creates it again. Use with caution!
+     */
+    fun replace(properties: Map<String, Any?>): RepositoryResult {
+        delete()
+        return save(properties)
+    }
+
+    /**
+     * Synchronizes on demand previously loaded properties of node (by default properties are loaded lazily).
```

Comment:
lazily because no need to read properties when removing node only

Reply:
I like lazy :)